driver/src/price_estimation/average_price_source.rs

```diff
@@ -1,41 +1,44 @@
 use super::{PriceSource, Token};
 use crate::models::TokenId;
 use anyhow::{anyhow, Result};
-use crossbeam_utils::thread;
+use futures::future::{self, BoxFuture, FutureExt as _};
 use std::collections::HashMap;
-use std::sync::Mutex;
 
 /// Combines two prices into one average.
 pub struct AveragePriceSource<T0, T1> {
-    source_0: Mutex<T0>,
-    source_1: Mutex<T1>,
+    source_0: T0,
+    source_1: T1,
 }
 
 impl<T0, T1> AveragePriceSource<T0, T1> {
     pub fn new(source_0: T0, source_1: T1) -> Self {
-        Self {
-            source_0: Mutex::new(source_0),
-            source_1: Mutex::new(source_1),
-        }
+        Self { source_0, source_1 }
     }
 }
 
-impl<T0: PriceSource + Send, T1: PriceSource + Send> PriceSource for AveragePriceSource<T0, T1> {
-    fn get_prices(&self, tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
-        let prices = thread::scope(|s| {
-            let handle_0 = s.spawn(|_| self.source_0.lock().unwrap().get_prices(tokens));
-            let handle_1 = s.spawn(|_| self.source_1.lock().unwrap().get_prices(tokens));
-            (handle_0.join().unwrap(), handle_1.join().unwrap())
-        })
-        .unwrap();
-        match prices {
-            (Ok(p0), Ok(p1)) => Ok(average_prices(p0, p1)),
-            (Ok(p), Err(e)) | (Err(e), Ok(p)) => {
-                log::warn!("one price source failed: {}", e);
-                Ok(p)
+impl<T0: PriceSource + Send + Sync, T1: PriceSource + Send + Sync> PriceSource
+    for AveragePriceSource<T0, T1>
+{
+    fn get_prices<'a>(
+        &'a self,
+        tokens: &'a [Token],
+    ) -> BoxFuture<'a, Result<HashMap<TokenId, u128>>> {
+        async move {
+            let prices = future::join(
```

Comment:
Any reason you prefer `future::join(...).await` over `futures::join!(...)`?

Reply:
In general if a function can accomplish the same thing I prefer the function. In this case I'm not sure what the difference even is internally, except that the macro always does `.await` internally. https://github.com/rust-lang/futures-rs/issues/2127

Reply:
I think it also selects `join`, or `join3`, `join4`, etc. based on the number of arguments. So if you were to add a new future to join, then you would just have to add an additional parameter.